src/core/component.ts

```diff
@@ -80,7 +79,7 @@ class Component<P> extends React.PureComponent<P, IState> {
               ...nestedTemplates,
               ...this.findNestedTemplate(child)
             };
-            args.push(this._registerNestedOption(child) || child);
+            args.push(this._preprocessChild(child) || child);
```

Comment:
I suggest to add smth like `abstract _getChildren` instead of `_preprocessChild`:
```ts
const children = this._getChildren().
if (!!children) {
  args.push(children);
}
```
in the case of `Component` it would be 
```ts
...
 React.Children.forEach(this.props.children, (child: React.ReactElement<any>) => {
  nestedTemplates = {
    ...nestedTemplates,
    ...this.findNestedTemplate(child)
  };
  children.push(this._registerExtension(child) || this._registerNestedOption(child) || child)
...
  return children;
}
```

in the case of `ExtensionComponent`:
```ts
return this.props.children
```
This will disable templates discovering for extension components, but do we really need it for the validator?

Reply:
Talked in private - such refactoring shouldn't be included in this PR.